api/files.py

```diff
@@ -0,0 +1,140 @@
+import datetime
+import hashlib
+import logging
+import shutil
+import json
+import os
+
+from . import util
+
+log = logging.getLogger('scitran.api')
+
+
+class FileRequest(object):
+    """This class provides and interface for file uploads.
+    To perform an upload the client of the class should follow these steps:
+
+    1) initialize the request
+    2) save a temporary file
+    3) check identical
+    4) move the temporary file to its destination
+
+    The operations could be safely interleaved with other actions like permission checks or database updates.
+    """
+
+    def __init__(self, client_addr, filename, body, received_md5, metadata, tags):
+        self.client_addr = client_addr
+        self.filename = filename
+        self.body = body
+        self.received_md5 = received_md5
+        self.metadata = metadata
+        self.tags = tags
+        self.mimetype = util.guess_mimetype(filename)
+        self.filetype = util.guess_filetype(filename, self.mimetype)
+
+    def save_temp_file(self, tempdir_path):
+        self.tempdir_path = tempdir_path
+        success, duration = self._save_temp_file(self.tempdir_path)
+        if not success:
+            self.abort(400, 'Content-MD5 mismatch.')
```

Comment:
Good catch.
self.abort is not available here.

success is false only if there is a Content-MD5 mismatch, so I prefer to be specific in the message.